delegate-app/Main.hs

```diff
@@ -1,62 +1,336 @@
 -- Plutus imports
 import Plutus.V2.Ledger.Api (POSIXTime (..))
 
 -- HydraAuction imports
 
 import HydraAuction.Delegate (
+  ClientId,
+  ClientResponseScope (Broadcast),
   DelegateEvent (..),
-  DelegateInput (DelegateEvent),
   DelegateRunnerT,
-  delegateStep,
+  clientIsInScope,
+  delegateEventStep,
+  delegateFrontendRequestStep,
   execDelegateRunnerT,
  )
-import HydraAuction.Delegate.Interface (DelegateResponse (..))
+import HydraAuction.Delegate.Interface (
+  DelegateResponse (AuctionSet),
+  FrontendRequest,
+ )
+import HydraAuction.Delegate.Server (
+  DelegateError (FrontendNoParse),
+  DelegateServerConfig (
+    DelegateServerConfig,
+    host,
+    ping,
+    port,
+    tick
+  ),
+  DelegateServerLog (
+    DelegateError,
+    FrontendConnected,
+    FrontendInput,
+    QueueAuctionPhaseEvent,
+    Started,
+    ThreadEvent
+  ),
+  DelegateTracerT,
+  QueueAuctionPhaseEvent (ReceivedAuctionSet),
+  ServerAppT,
+  ThreadEvent (ThreadCancelled, ThreadStarted),
+  ThreadSort (DelegateRunnerThread, QueueAuctionStageThread, WebsocketThread),
+ )
 import HydraAuction.OnChain.Common (secondsLeftInInterval, stageToInterval)
 import HydraAuction.Tx.Common (currentAuctionStage, currentTimeMilliseconds)
+import HydraAuction.Types (AuctionTerms)
+import HydraAuctionUtils.Tracing (
+  MonadTracer (trace),
+  askTracer,
+  runWithTracer',
+ )
+
+{- | per websocket, a thread gets spawned which then
+   subscribes to the broadcast channel and sends to
+   the incoming channel of the delegate runner
+-}
+websocketsServer ::
+  -- | the amount of seconds between pings
+  Int ->
+  IO ClientId ->
+  -- | the queue for the frontend requests of the delegates
+  --   (this is where the requests are enqueued)
+  TQueue (ClientId, FrontendRequest) ->
+  -- | the queue for the broadcast (this is where we receives events from the runner)
+  TChan (ClientResponseScope, DelegateResponse) ->
+  ServerAppT (DelegateTracerT IO)
+websocketsServer
+  pingSecs
+  genFreshClientId
+  frontendRequestQueue
+  broadcast
+  pending = do
+    connection <- liftIO $ acceptRequest pending
+    clientId <- liftIO genFreshClientId
+
+    trace $ FrontendConnected clientId
+    tracer <- askTracer
+
+    liftIO $ do
+      toClientsChannelCopy <- atomically $ dupTChan broadcast
+      -- From the Documentation of 'withPingThread' in 'Network.Websockets':
+      -- This is useful to keep idle connections open through proxies and whatnot.
+      -- Many (but not all) proxies have a 60 second default timeout, so based on
+      -- that sending a ping every 30 seconds is a good idea.
+      withPingThread connection pingSecs (pure ()) $
+        race_
+          (forever $ receiveAndPutInQueue tracer connection clientId)
+          ( forever $
+              sendFromChannel connection clientId toClientsChannelCopy
+          )
+    where
+      sendToClient :: forall a. ToJSON a => Connection -> a -> IO ()
+      sendToClient connection = sendTextData connection . encode
```

Comment:
this name has less information than before, `sendTextData` also sends to client, I don't know how this is better than `encodeSend` which at least expresses all it does.

Reply:
All local functions send something to queue or to client. So while I was reading this code I tried to understand which part send where.

And encoding to text or to bytecode is implementation detail which is only important to client, and does not matter for design of this module.